app/app.go

```diff
@@ -308,3 +313,26 @@ func (a *app) submitFileTransferAuditlog(ctx context.Context, userID string, dev
 	}
 	return nil
 }
+
+func (a *app) Shutdown(ctx context.Context, timeout time.Duration) error {
+	nCancels := len(a.shutdownCancels)
+	reducedTimeout := timeout - 5*time.Second
+	for _, cancel := range a.shutdownCancels {
+		cancel()
+		time.Sleep(reducedTimeout / time.Duration(nCancels))
+	}
+	return nil
+}
+
+func (a *app) RegisterShutdownCancel(cancel context.CancelFunc) string {
+	if uuid, err := uuid.NewRandom(); err != nil {
+		id := uuid.String()
+		a.shutdownCancels[id] = cancel
+		return id
+	}
+	return ""
```

Comment:
To avoid having to generate a random ID and make it slightly more memory efficient:
```suggestion
var shutdownID uint32

func (a *app) RegisterShutdownCancel(cancel context.CancelFunc) uint32 {
	id := atomic.AddUint32(&shutdownID, 1)
	a.shutdownCancels[id] = cancel
	return id
```